File: cs/ccxt/ws/Stream.cs
namespace ccxt.pro;

using System;
using System.Collections.Generic;
using System.Linq;

public class Stream : IBaseStream
{
    public int maxMessagesPerTopic { get; set; }

    public bool verbose { get; set; }
    private Dictionary<string, List<Message>> topics;
    private Dictionary<string, List<Consumer>> consumers;
    public List<Dictionary<string, object>> activeWatchFunctions;
    private Dictionary<string, int> topicIndexes;
    public Stream(int? maxMessagesPerTopic = null, bool? verbose = null)
    {
        Init(maxMessagesPerTopic, verbose);
    }

    private void Init(int? maxMessagesPerTopic = null, bool? verbose = null)
    {
        this.maxMessagesPerTopic = maxMessagesPerTopic ?? 0;
        this.verbose = verbose ?? false;
        this.topics = new Dictionary<string, List<Message>>();
        this.consumers = new Dictionary<string, List<Consumer>>();
        this.activeWatchFunctions = new List<Dictionary<string, object>>();
        this.topicIndexes = new Dictionary<string, int>();
    }

    public void produce(object topic2, object payload, object error = null)
    {
        string topic = topic2 as String;
        if (!topics.ContainsKey(topic))
        {
            topics[topic] = new List<Message>();
        }
        if (!topicIndexes.ContainsKey(topic))
        {
            topicIndexes[topic] = -1;
        }
        topicIndexes[topic] += 1;
        var index = topicIndexes[topic];
        var messages = topics[topic];

        var message = new Message
        {
            payload = payload,
            error = error,
            metadata = new Metadata
            {
                stream = this,
                topic = topic,
                index = index,
                history = messages.ToList(), // Creates a shallow copy
            },
        };

        if (messages.Count > maxMessagesPerTopic)
        {
            messages.RemoveAt(0);
        }
        if (maxMessagesPerTopic != 0)
        {
            messages.Add(message);
        }

        if (consumers.ContainsKey(topic))
        {
            foreach (var consumer in consumers[topic])
            {
                consumer.publish(message);
            }
        }
        if (this.verbose)
        {
            Console.WriteLine($"Published message to topic: {topic}");
        }
    }

    public void subscribe(object topic2, object consumerFn2, object params2 = null)
    {
        var topic = topic2 as String;
        var paramsDict = params2 as Dictionary<string, object> ?? new Dictionary<string, object>();
        var synchronous = paramsDict.ContainsKey("synchronous") ? (bool)paramsDict["synchronous"] : true;
        var consumerMaxBacklogSize = paramsDict.ContainsKey("consumerMaxBacklogSize") ? (int)paramsDict["consumerMaxBacklogSize"] : 1000;

        // Check if consumerFn2 is already a ConsumerFunction
        ConsumerFunction consumerFn = consumerFn2 as ConsumerFunction;

        // If it's not a ConsumerFunction, check if it's a Func<Message, Task>
        if (consumerFn == null && consumerFn2 is Func<Message, Task> func)
        {
            // Convert Func<Message, Task> to ConsumerFunction
            consumerFn = new ConsumerFunction(func.Invoke);
        }
        else if (consumerFn == null && consumerFn2 is Action<Message> action)
        {
            // Convert Action<Message> (void-returning) to ConsumerFunction
            consumerFn = new ConsumerFunction((msg) => { action.Invoke(msg); return Task.CompletedTask; });
        }
        if (consumerFn == null)
        {
            Console.WriteLine("Consumer function is required");
            throw new Exception("Consumer function is required");
        }
        var consumer = new Consumer(consumerFn, GetLastIndex(topic), new ConsumerOptions 
        { 
            synchronous = synchronous,
            maxBacklogSize = consumerMaxBacklogSize
        });

        if (!consumers.ContainsKey(topic))
        {
            consumers[topic] = new List<Consumer>();
        }

        consumers[topic].Add(consumer);
        if (this.verbose)
        {
            Console.WriteLine($"Subscribed Consumer {consumerFn.Method.Name} to {topic}, synchronous: {synchronous}, maxBacklogSize: {consumerMaxBacklogSize}");
        }
    }

    public bool unsubscribe(object topic2, ConsumerFunction consumerFn)
    {
        string topic = topic2 as String;
        if (consumers.ContainsKey(topic))
        {
            consumers[topic] = consumers[topic].Where(consumer => !consumer.fn.Equals(consumerFn)).ToList();
            if (this.verbose)
            {
                Console.WriteLine($"Unsubscribed {consumerFn.Method.Name} from {topic}.");
            }
            return true;
        }
        else
        {
            if (this.verbose)
            {
                Console.WriteLine($"Unable to unsubscribe {consumerFn.Method.Name} from {topic}. Consumer not found.");
            }
            return false;
        }
    }

    public List<Message> GetMessageHistory(string topic)
    {
        return topics.ContainsKey(topic) ? topics[topic] : new List<Message>();
    }

    public int GetLastIndex(string topic)
    {
        if (topicIndexes.ContainsKey(topic))
        {
            return topicIndexes[topic];
        }
        return -1;
    }

    public void addWatchFunction(string watchFn, List<object> args)
    {
        this.activeWatchFunctions.Add(new Dictionary<string, object> { { "method", watchFn }, { "args", args } });
    }

    public void close()
    {
        if (this.verbose)
        {
            Console.WriteLine("Closed Stream");
        }
        Init(maxMessagesPerTopic);
    }
}

File: php/pro/Stream.php
<?php

namespace ccxt\pro;

use ccxt\pro\Consumer;

class Metadata {
    public $stream;
    public $topic;
    public $index;
    public $history;

    public function __construct($stream, $topic, $index) {
        $this->stream = $stream;
        $this->topic = $topic;
        $this->index = $index;
        $this->history = $this->stream->get_message_history($topic);
    }
}

class Message {
    public $payload;
    public $error;
    public $metadata;

    public function __construct($payload, $error, $stream, $topic, $index) {
        $this->payload = $payload;
        $this->error = $error;
        $this->metadata = new Metadata($stream, $topic, $index);
    }
}

class Stream {
    public $max_messages_per_topic;
    public $verbose;
    public $topics = array();
    private $consumers = array();
    public $active_watch_functions = array();
    public $topicIndexes = array();

    public function __construct($max_messages_per_topic = 0, $verbose = false) {
        $this->max_messages_per_topic = $max_messages_per_topic;
        $this->verbose = $verbose;
        $this->topicIndexes = array();
    }

    public function produce($topic, $payload, $error = null) {
        if (!isset($this->topics[$topic])) {
            $this->topics[$topic] = [];
        }
        if (!isset($this->topicIndexes[$topic])) {
            $this->topicIndexes[$topic] = -1;
        }
        $this->topicIndexes[$topic] += 1;
        $index = $this->topicIndexes[$topic];
        $messages = $this->topics[$topic];
        $message = new Message($payload, $error, $this, $topic, $index);

        if (count($messages) > $this->max_messages_per_topic) {
            array_shift($messages);
        }
        if ($this->max_messages_per_topic !== 0) {
            array_push($this->topics[$topic], $message);
        }

        $consumers = $this->consumers[$topic] ?? [];
        $this->send_to_consumers($consumers, $message);
    }

    public function subscribe($topic, callable $consumerFn, $synchronous = true) {
        $consumer = new Consumer($consumerFn, $this->get_last_index($topic), ['synchronous' => $synchronous]);
        if (!isset($this->consumers[$topic])) {
            $this->consumers[$topic] = [];
        }
        $this->consumers[$topic][] = $consumer;
        if ($this->verbose) {
            print_r ("Subscribed function to topic: " . $topic . "\n");
        }
    }

    public function unsubscribe($topic, callable $consumerFn) {
        if (isset($this->consumers[$topic])) {
            $this->consumers[$topic] = array_filter($this->consumers[$topic], function ($consumer) use ($consumerFn) {
                return $consumer->fn !== $consumerFn;
            });
            if ($this->verbose) {
                print_r ("Unsubscribed function from topic: " . $topic . "\n");
            }
            return true;
        }
        if ($this->verbose) {
            print_r ("Unable to unsubscribe function from topic: " . $topic . ". Consumer function not found.\n");
        }
        return false;
    }

    public function get_message_history($topic) {
        return isset($this->topics[$topic]) ? $this->topics[$topic] : [];
    }

    private function get_last_index($topic): int {
        if (isset($this->topicIndexes[$topic])) {
            return $this->topicIndexes[$topic];
        }
        return -1;
    }

    private function send_to_consumers($consumers, $message) {
        if ($this->verbose) {
            print_r ('sending message from topic ' . $message->metadata->topic, 'to consumers');
        }
        foreach ($consumers as $consumer) {
            $consumer->publish($message);
        }
    }

    public function add_watch_function($watchFn, $args) {
        $this->active_watch_functions[] = array('method' => $watchFn, 'args' => $args);
    }

    
    public function close () {
        $this->topics = [];
        $this->consumers = [];
        $this->topicIndexes = array();
        if ($this->verbose) {
            print_r ("Closed stream\n");
        }
    }
}
